keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+
+#### Story 2
+
+A Kubernetes developer would like use a new great etcd feature, but it is not available on all supported etcd versions. What can they do?
+Historically, Kubernetes avoided using any features that were not available on all supported versions of etcd.
+From K8s 1.31, Kubernetes added very complicated and fragile system: [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to detect etcd version, parse it and guess whether this feature based on a set of hardcoded etcd version. It does not really know if the feature is enabled by the feature `--experimental` flag.
+
+With a client-side mechanism to reliably determine if a cluster-wide feature is enabled,  would allow Kubernetes to immediately utilize new features.
+
+#### Story 3
+
+In a HA cluster, users would like to turn off an enabled feature. They need to restart each server node one by one with feature changed from enabled to disabled. Today, after the restarting process begins and before all nodes are restarted, the behavior of the feature is really undefined: it is enabled on some nodes and disabled on the other nodes.
+
+Now with the cluster feature gate, we can deterministically tell if the feature is enabled or disabled for the whole cluster at any moment.
+
+This story applies similar to the upgrade/downgrade scenarios as well.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Could it done on the client side instead of inside etcd? 
+One could argue that it is easy to know if a feature enabled for the whole cluster by querying each member if the feature is enabled on the client side. But there are several caveates with this approach:
+* not all members are available all the time. if some member is not available, it is questionable if the feature can be used, and that could break the HA assumption. 
+* some feature might change how a raft message is sent and applied, and the order of messages relative to the index the feature is enabled is critical. Here the consistent index is more important than the time, and it is hard to get that outside etcd. 
+
+#### Should we prevent a new member with an inconsistent feature gate from joining the cluster?
+
+One way to prevent inconsistencies of feature gate between different members is to prevent a new member with an inconsistent feature gate from joining the cluster in the first place. 
+We do not think this is the right way because we want to keep the ability to turn on/off a feature in a rolling sequence without bringing down the whole cluster. 
+
+
+### Risks and Mitigations
+
+#### Backport Risks
+
+Because we are adding a new type of raft request, if we want to support feature gate as soon as in 3.6, we have to backport the proto changes and add no-op apply logic in 3.5. 
```

Comment:
Not sure I understand the risk, it should be mitigated by not using the new proto if the ClusterVersion is low.

Reply:
I am thinking about sending the `proposed_feature_gates` in the `ClusterMemberAttrSetRequest`, which is sent to raft right after start, before the cluster version is known. So there is a chance it is sent to a lower version member. 

If we enforce not using the new proto if the ClusterVersion is low, there would be a few places to add the ugly code of `if s.ClusterVersion() < "3.6"`, and `ClusterMemberAttrSetRequest` needs to be updated whenever cluster version changes. 
Backporting the proto seems to be cleaner and less error prone.

Reply:
removed.